File: common/src/main/java/dev/obscuria/fragmentum/service/NetworkService.java
package dev.obscuria.fragmentum.service;

import dev.obscuria.fragmentum.v2.api.common.network.PayloadRegistrar;
import net.minecraft.core.BlockPos;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;

public interface NetworkService {

    PayloadRegistrar registrar(String modId);

    <T extends CustomPacketPayload> void reply(T payload);

    <T extends CustomPacketPayload> void sendTo(ServerPlayer player, T payload);

    <T extends CustomPacketPayload> void sendToAllTracking(ServerLevel level, BlockPos pos, T payload);

    <T extends CustomPacketPayload> void sendToAllTracking(Entity entity, T payload);

    <T extends CustomPacketPayload> void sendToAll(MinecraftServer server, T payload);

    <T extends CustomPacketPayload> void sendToServer(T payload);
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/api/service/NetworkService.kt
package dev.obscuria.fragmentum.api.service

import dev.obscuria.fragmentum.api.network.PayloadRegistrar
import net.minecraft.core.BlockPos
import net.minecraft.server.MinecraftServer
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity

interface NetworkService
{
    fun payloadRegistrar(modId: String) : PayloadRegistrar

    fun <T : Any> reply(payload: T)

    fun <T : Any> sendTo(player: ServerPlayer, payload: T)

    fun <T : Any> sendToTracking(world: ServerLevel, pos: BlockPos, payload: T)

    fun <T : Any> sendToTracking(entity: Entity, payload: T)

    fun <T : Any> sendToAll(server: MinecraftServer, payload: T)

    fun <T : Any> sendToServer(payload: T)
}